addons/l10n_it_edi_doi/models/declaration_of_intent.py

```diff
@@ -0,0 +1,309 @@
+    )
+
+    not_yet_invoiced = fields.Monetary(
+        string='Not Yet Invoiced',
+        compute='_compute_not_yet_invoiced',
+        store=True,
+        readonly=True,
+        help="Total amount of planned sales under this Declaration of Intent (i.e. current quotation and sales orders) that can still be invoiced",
+    )
+
+    remaining = fields.Monetary(
+        string='Remaining',
+        compute='_compute_remaining',
+        store=True,
+        readonly=True,
+        help="Remaining amount after deduction of the Invoiced and Not Yet Invoiced amounts.",
+    )
+
+    protocol_number_part1 = fields.Char(
+        string='Protocol 1',
+        required=True,
+        readonly=False,
+        copy=False,
+    )
+
+    protocol_number_part2 = fields.Char(
+        string='Protocol 2',
+        required=True,
+        readonly=False,
+        copy=False,
+    )
+
+    invoice_ids = fields.One2many(
+        'account.move',
+        'l10n_it_edi_doi_id',
+        string="Invoices / Refunds",
+        copy=False,
+        readonly=True,
+    )
+
+    sale_order_ids = fields.One2many(
+        'sale.order',
+        'l10n_it_edi_doi_id',
+        string="Sales Orders / Quotations",
+        copy=False,
+        readonly=True,
+    )
+
+    _sql_constraints = [
+        ('protocol_number_unique',
+         'unique(protocol_number_part1, protocol_number_part2)',
+         "The Protocol Number of a Declaration of Intent must be unique! Please choose another one."),
+        ('threshold_positive',
+         'CHECK(threshold > 0)',
+         "The Threshold of a Declaration of Intent must be positive."),
+    ]
+
+    @api.depends('protocol_number_part1', 'protocol_number_part2')
+    def _compute_display_name(self):
+        for record in self:
+            record.display_name = f"{record.protocol_number_part1}-{record.protocol_number_part2}"
+
+    @api.depends('invoice_ids', 'invoice_ids.state', 'invoice_ids.l10n_it_edi_doi_amount')
+    def _compute_invoiced(self):
+        for declaration in self:
+            relevant_invoices = declaration.invoice_ids.filtered(
+                lambda invoice: invoice.state == 'posted'
+            )
+            declaration.invoiced = sum(relevant_invoices.mapped('l10n_it_edi_doi_amount'))
+
+    @api.depends('sale_order_ids', 'sale_order_ids.state', 'sale_order_ids.l10n_it_edi_doi_not_yet_invoiced')
+    def _compute_not_yet_invoiced(self):
+        for declaration in self:
+            relevant_orders = declaration.sale_order_ids.filtered(
+                lambda order: order.state == 'sale'
+            )
+            declaration.not_yet_invoiced = sum(relevant_orders.mapped('l10n_it_edi_doi_not_yet_invoiced'))
+
+    @api.depends('threshold', 'not_yet_invoiced', 'invoiced')
+    def _compute_remaining(self):
+        for record in self:
+            record.remaining = record.threshold - record.invoiced - record.not_yet_invoiced
```

Comment:
I think it makes sense since that we we can easily see that sth. went wrong with the declaration and by how much.
Else declarations that hit the threshold exactly look like exceeded declarations?
And the more they are exceeded the more problematic they are I guess.